bee-network/bee-autopeering/src/peering/neighbor.rs

```diff
@@ -0,0 +1,347 @@
+    pub(crate) fn is_full(&self) -> bool {
+        self.read().is_full()
+    }
+
+    /// Collect all peers belonging to the neighborhood into a `Vec`.
+    pub(crate) fn peers(&self) -> Vec<Peer> {
+        self.read().neighbors.iter().map(|d| d.peer()).cloned().collect()
+    }
+
+    fn read(&self) -> RwLockReadGuard<NeighborhoodInner<N, INBOUND>> {
+        self.inner.read().expect("error getting read access")
+    }
+
+    fn write(&self) -> RwLockWriteGuard<NeighborhoodInner<N, INBOUND>> {
+        self.inner.write().expect("error getting write access")
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct NeighborhoodInner<const N: usize, const INBOUND: bool> {
+    neighbors: Vec<Neighbor>,
+}
+
+impl<const N: usize, const INBOUND: bool> NeighborhoodInner<N, INBOUND> {
+    fn insert_neighbor(&mut self, peer: Peer, local: &Local) -> bool {
+        // If the peer already exists remove it.
+        // NOTE: It's a bit less efficient doing it like this, but the code requires less branching this way.
+        let _ = self.remove_neighbor(peer.peer_id());
+
+        if self.neighbors.len() >= N {
+            return false;
+        }
+
+        // Calculate the distance to that peer.
+        let distance = salt_distance(&local.peer_id(), peer.peer_id(), &{
+            if INBOUND {
+                local.private_salt().expect("missing private salt")
+            } else {
+                local.public_salt().expect("missing public salt")
+            }
+        });
+
+        self.neighbors.push(Neighbor { distance, peer });
+
+        true
+    }
+
+    fn remove_neighbor(&mut self, peer_id: &PeerId) -> Option<Peer> {
+        if self.neighbors.is_empty() {
+            None
+        } else if let Some(index) = self.neighbors.iter().position(|pd| pd.peer().peer_id() == peer_id) {
+            let Neighbor { peer, .. } = self.neighbors.remove(index);
+            Some(peer)
+        } else {
+            None
+        }
+    }
+
+    fn is_preferred(&mut self, candidate: &Neighbor) -> bool {
+        if let Some(furthest) = self.find_furthest() {
+            candidate < furthest
+        } else {
+            true
+        }
+    }
+
+    fn select_from_candidate_list<'a>(&mut self, candidates: &'a [&'a Neighbor]) -> Option<&'a Peer> {
+        if candidates.is_empty() {
+            None
+        } else if let Some(furthest) = self.find_furthest() {
+            for candidate in candidates {
+                if *candidate < furthest {
+                    return Some(candidate.peer());
+                }
+            }
+            None
+        } else {
+            // Any candidate can be selected: pick the first.
+            Some(candidates[0].peer())
+        }
+    }
+
+    fn find_furthest(&mut self) -> Option<&Neighbor> {
+        if self.neighbors.len() >= N {
+            self.neighbors.sort_unstable();
+            self.neighbors.last()
+        } else {
+            None
```

Comment:
Is there no furthest if it's not full ?

Reply:
yes, finding the furthest away neighbor makes only sense in a full neighborhood, because if it's not full, then ANY neighbor is good enough.

I renamed the method to `find_furthest_if_full` which should hopefully make this clear.

Reply:
If we don't want to do that, then we need two separate API calls `find_furthest` and `is_full`, which then would be better in terms of understanding. Let's maybe consider this in a patch version?!